rsp/runtimes/bundles/org.jboss.tools.rsp.server.generic/src/main/java/org/jboss/tools/rsp/server/generic/servertype/actions/ShowInBrowserActionHandler.java

```diff
@@ -0,0 +1,157 @@
+		
+		List<String> deployments2 = new ArrayList<>();
+		deployments2.add(ACTION_SHOW_IN_BROWSER_SELECT_SERVER_ROOT);
+		deployments2.addAll(deployments);
+		
+		prompt.setValidResponses(deployments2);
+		item1.setPrompt(prompt);
+		
+		items.add(item1);
+		workflow.setItems(items);
+		return action;
+	}
+	
+	private List<DeployableState> getDeployableStatesHavingContextRoots() {
+		List<DeployableState> dss = getDeployableStates();
+		ArrayList<DeployableState> collector = new ArrayList<>();
+		for( DeployableState ds : dss ) {
+			if( getContextRoot(ds) != null ) 
+				collector.add(ds);
+		}
+		return collector;
+	}
+	
+	protected List<DeployableState> getDeployableStates() {
+		return genericServerBehavior.getServerPublishModel().getDeployableStatesWithOptions();
+	}
+
+	private String getContextRoot(DeployableState ds) {
+		DeployableReference ref = ds.getReference();
+		String outputName = getOutputName(ref);
+		
+		if( outputName.endsWith(".war") || outputName.endsWith(".ear")) {
+			return outputName.substring(0, outputName.length() - 4);
+		}
+		return null;
+	}
+	
+	protected String getOutputName(DeployableReference ref) {
+		Map<String, Object> options = ref.getOptions();
+		String def = null;
+		if( ref.getPath() != null ) {
+			def = new File(ref.getPath()).getName();
+		}
+		String k = ServerManagementAPIConstants.DEPLOYMENT_OPTION_OUTPUT_NAME; 
+		if( options != null && options.get(k) != null ) {
+			return (String)options.get(k);
+		}
+		return def;
+	}
+	
+	public WorkflowResponse handle(ServerActionRequest req) {
+		if( req == null || req.getData() == null ) 
+			return AbstractServerDelegate.cancelWorkflowResponse();
+			
+		String choice = (String)req.getData().get(ACTION_SHOW_IN_BROWSER_SELECTED_PROMPT_ID);
+		if( choice == null )
+			return AbstractServerDelegate.cancelWorkflowResponse();
+		
+		String url = findUrlFromChoice(choice);
+		if( url != null ) {
+			WorkflowResponseItem item = new WorkflowResponseItem();
+			item.setItemType(ServerManagementAPIConstants.WORKFLOW_TYPE_OPEN_BROWSER);
+			item.setLabel("Open the following url: " + url);
+			item.setContent(url);
+			WorkflowResponse resp = new WorkflowResponse();
+			resp.setItems(Arrays.asList(item));
+			resp.setStatus(StatusConverter.convert(Status.OK_STATUS));
+			return resp;
+		}
+		return AbstractServerDelegate.cancelWorkflowResponse();
+	}
+	
+	private String findUrlFromChoice(String choice) {
+		String baseUrl = getBaseUrl();
+		if( choice.equals(ACTION_SHOW_IN_BROWSER_SELECT_SERVER_ROOT)) {
+			return baseUrl;
+		} else {
+			List<DeployableState> states = getDeployableStates();
+			for( DeployableState ds : states ) {
+				if( ds.getReference().getPath().equals(choice)) {
+					String contextRoot = getContextRoot(ds);
+					if( contextRoot != null ) {
+						return baseUrl + "/" + contextRoot;
+					}
+				}
+			}
+		}
+		return null;
+	}
+	
+	protected String getBaseUrl() {
+		return genericServerBehavior.getPollURL(genericServerBehavior.getServer());
```

Comment:
There's no guarantee a specific server is using web port poller and has that property set. It'd be better if the json for the action here had its own property listing the url. Yeah it leads to some duplication, but it's better than a possibly nonsensical answer.